src/MongoDB.Driver.Core/Core/Servers/TopologyVersion.cs

```diff
@@ -0,0 +1,326 @@
+        }
+
+#if NET452 || NETSTANDARD2_0
+        private TopologyVersion(SerializationInfo info, StreamingContext context)
+        {
+            _processId = (ObjectId) info.GetValue("_processId", typeof(ObjectId));
+            _counter = (long) info.GetValue("_counter", typeof(long));
+            _hashCode = new Hasher().Hash(_processId).Hash(_counter).GetHashCode();
+        }
+#endif
+
+        // properties
+        /// <summary>
+        /// Gets the process identifier.
+        /// </summary>
+        /// <value>
+        /// The process identifier.
+        /// </value>
+        public ObjectId ProcessId => _processId;
+
+        /// <summary>
+        /// Gets the counter.
+        /// </summary>
+        /// <value>
+        /// The counter.
+        /// </value>
+        public long Counter => _counter;
+
+        // methods
+        /// <summary>
+        /// Gets whether <paramref name="x"/>.Equals(<paramref name="y"/>).
+        /// </summary>
+        /// <param name="x">A TopologyVersion.</param>
+        /// <param name="y">A TopologyVersion.</param>
+        /// <returns>
+        /// Whether <paramref name="x"/>.Equals(<paramref name="y"/>).
+        /// </returns>
+        public static bool operator ==(TopologyVersion x, TopologyVersion y) => CompareFreshnessOfLocalToServerResponse(x, y) == 0;
+
+        /// <summary>
+        /// Gets whether <paramref name="x"/>  != (<paramref name="y"/>).
+        /// </summary>
+        /// <param name="x">A TopologyVersion.</param>
+        /// <param name="y">A TopologyVersion.</param>
+        /// <returns>
+        /// Whether <paramref name="x"/> != (<paramref name="y"/>).
+        /// </returns>
+        public static bool operator !=(TopologyVersion x, TopologyVersion y) => CompareFreshnessOfLocalToServerResponse(x, y) != 0;
+
+        /// <summary>
+        /// Compares a local TopologyVersion with aserver's TopologyVersion and indicates whether the local
+        /// TopologyVersion is staler, fresher, or equal to the server's TopologyVersion.
+        /// Per the SDAM specification, if the ProcessIds are not equal, this method assumes that
+        /// <paramref name="serverResponse"/> is more recent. This means that this method does not exhibit
+        /// the reversal properties of inequalities i.e. a "&lt;" b does not imply b "&gt;" a.
+        /// </summary>
+        /// <param name="local"> The locally saved TopologyVersion. </param>
+        /// <param name="serverResponse">The TopologyVersion received from a server.</param>
+        /// <returns>
+        /// Less than zero indicates that the local description is older than the response.
+        /// Zero indicates that the local description is equal to the response.
+        /// Greater than zero indicates that the local description is newer than the response.
+        /// </returns>
+        public static int CompareFreshnessOfLocalToServerResponse(TopologyVersion local, TopologyVersion serverResponse)
```

Comment:
So, IMHO, comparing TopologyVersions is a very error prone process as much of "intuition" here for CompareTo or the inequality operators doesn't apply particularly well. That being said, I've tried to simplify the logic by splitting it up into two parts. The first part, here, compares `TopologyVersion`s, which because they are structs, cannot be null. The second part, in the extension methods below, compares `Nullable<TopologyVersion>`s, which CAN be null and has additional logic. It may be helpful to think of `TopologyVersion` as a "base class/struct" and `Nullable<TopologyVersion>` as a "subclass/substruct" with overridden "CompareTo" logic.